### PR TITLE
SIA-R10: Accept `webauthn` token

### LINKED ISSUE #1332: SIA-R10 should accept `webauthn` token
See https://github.com/act-rules/act-rules.github.io/pull/2017

### PR DESCRIPTION
**Pull request by rcj-siteimprove**
Resolves #1332

The rule has been updated to accept autocomplete attribute token `webauthn`.
In addition a few bugs where discovered and fixed. The bugs where:
* Attributes with modifiers like `home` followed by nothing were incorrectly passing
* Attributes with valid tokens followed by invalid tokens were incorrectly passing

### CODE REVIEW

File: packages/alfa-rules/src/sia-r10/rule.ts [hunk 1]
```diff
@@ -76,105 +77,116 @@ function hasTokens(input: string): boolean {
...
-  if (field === null) {
-    return false;
+  // 3. Either of the following two options:
+  if (unmodifiableFields.includes(next)) {
+    next = tokens[i++];
+  } else if (modifiers.includes(next)) {
+    next = tokens[i++];
+    // Optional, but must be followed by:
+    if (modifiableFields.includes(next)) {
+      next = tokens[i++];
```
Comment (Jym77) [inline packages/alfa-rules/src/sia-r10/rule.ts:114]: Looks like the modifier is here mandatory before a modifiable field, but shouldn't.
The grammar here should be `unmodifiable | [modifier]? modifiable`, but the code seems to accept `unmodifiable | modifier modifiable`.

File: packages/alfa-rules/src/sia-r10/rule.ts [hunk 2]
```diff
@@ -76,105 +77,116 @@ function hasTokens(input: string): boolean {
   return input.trim() !== "" && input.split(/\s+/).length > 0;
 }
 
+/**
+ * {@link https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#autofill-detail-tokens}
```
Comment (Jym77) [inline packages/alfa-rules/src/sia-r10/rule.ts:81]:
```suggestion
 * {@link https://html.spec.whatwg.org/multipage/#autofill-detail-tokens}
```
(shortening URL a bit, and depending less on the page structure that might change)

File: packages/alfa-rules/test/sia-r10/rule.spec.tsx [hunk 3]
```diff
@@ -32,6 +32,84 @@ test("evaluate() passes a valid simple autocomplete attribute on an <input> elem
   ]);
 });
 
+test("evaluate() passes autocomplete attribute with \`section-blue shipping\`", async (t) => {
+  const element = <input autocomplete="section-blue shipping" />;
```
Comment (Jym77) [inline packages/alfa-rules/test/sia-r10/rule.spec.tsx:36]: That should fail due to no Step 3 "either of the following to options".

File: packages/alfa-rules/test/sia-r10/rule.spec.tsx [hunk 4]
```diff
@@ -32,6 +32,84 @@ test("evaluate() passes a valid simple autocomplete attribute on an <input> elem
...
+
+  t.deepEqual(await evaluate(R10, { document }), [
+    passed(R10, target, {
+      1: Outcomes.HasValidValue,
+    }),
+  ]);
+});
+
+test("evaluate() passes autocomplete attribute with \`webauthn\` token", async (t) => {
+  const element = <input autocomplete="webauthn" />;
```
Comment (Jym77) [inline packages/alfa-rules/test/sia-r10/rule.spec.tsx:49]: That should also fail due to the lack of mandatory field.

**Review by Jym77 — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/test/sia-r10/rule.spec.tsx [hunk 5]
```diff
@@ -58,6 +97,45 @@ test("evaluate() fails an autocomplete attribute with terms in wrong order", asy
...
+  const document = h.document([element]);
+
+  t.deepEqual(await evaluate(R10, { document }), [
+    failed(R10, target, {
+      1: Outcomes.HasNoValidValue,
+    }),
+  ]);
+});
+
+test("evaluate() fails an autocomplete attribute with \`home\` followed by nothing", async (t) => {
```
Comment (rcj-siteimprove) [inline packages/alfa-rules/test/sia-r10/rule.spec.tsx:113]:
```suggestion
test("evaluate() fails an autocomplete attribute with modifier followed by nothing", async (t) => {
```

File: packages/alfa-rules/test/sia-r10/rule.spec.tsx [hunk 6]
```diff
@@ -58,6 +97,45 @@ test("evaluate() fails an autocomplete attribute with terms in wrong order", asy
...
+  const document = h.document([element]);
+
+  t.deepEqual(await evaluate(R10, { document }), [
+    failed(R10, target, {
+      1: Outcomes.HasNoValidValue,
+    }),
+  ]);
+});
+
+test("evaluate() fails an autocomplete attribute with \`home webauthn\`", async (t) => {
```
Comment (rcj-siteimprove) [inline packages/alfa-rules/test/sia-r10/rule.spec.tsx:126]:
```suggestion
test("evaluate() fails an autocomplete attribute with modifier token followed by invalid token", async (t) => {
```

**Review by Jym77 — APPROVED** [review]

**Pull request merged**